### PR TITLE
SINDy-PI

### PR DESCRIPTION
**Pull request by akaptano**
Implemented the SINDy-PI optimization. Require a new library routine that is somewhat involved, and a new optimizer that relies on CVXPy. A simple example notebook solves an easy 1D implicit problem, will probably make this more exciting later. There is also a small change in model.print() for PySINDy, to correctly output the implicit models.

### CODE REVIEW

File: pysindy/optimizers/sindymodified_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        self.differentiation_method = differentiation_method
+        self.thresholds = thresholds
+        self.reg = get_regularization(thresholder)
+        self.unbias = False
+        self.optimization_method = optimization_method
+        self.learning_rate = learning_rate
+        self.soft_start = soft_start
+        self.q = q
+        self.c = c
+        self.t = t
```
Comment (briandesilva) [inline pysindy/optimizers/sindymodified_optimizer.py:161]: Are there more descriptive names we can use for these variables or are you following what's used in the paper?

File: pysindy/pysindy.py [hunk 2]
```diff
@@ -398,11 +400,20 @@ def print(self, lhs=None, precision=3):
...
-            elif lhs is None:
-                print(self.feature_names[i] + "' = " + eqn)
+                names = "(" + feature_names[i] + ")"
+                print(names + "[k+1] = " + eqn)
+            elif lhs is None and not isinstance(self.feature_library, SINDyPILibrary):
+                names = "(" + feature_names[i] + ")"
+                print(names + "' = " + eqn)
+            elif lhs is None and isinstance(self.feature_library, SINDyPILibrary):
+                names = feature_names[i]
+                print(names + " = " + eqn)
```
Comment (briandesilva) [inline pysindy/pysindy.py:416]: This would be clearer with nested conditionals, e.g.
```python
elif lhs is None:
    if isinstance(self.feature_library, SINDyPILibrary):
        ...
    else:
        ...
```

File: pysindy/pysindy.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 from numpy import vstack
 from numpy import zeros
 from scipy.integrate import odeint
+from scipy.integrate import solve_ivp
```
Comment (briandesilva) [inline pysindy/pysindy.py:11]: What's the motivation behind adding `solve_ivp`?

File: pysindy/utils/ODEs.py [hunk 4]
```diff
@@ -0,0 +1,501 @@
+import numpy as np
```
Comment (briandesilva) [inline pysindy/utils/ODEs.py:1]: Nice add! The filename should be lowercased "odes.py" for consistency.

**Review by briandesilva — CHANGES_REQUESTED** [review]
This is an initial review. I still need to look over (1) the SINDy-PI optimizer, the tests, and the updates to the notebooks. I just figured I'd get you some comments sooner rather than later.

File: pysindy/feature_library/SINDyPI_library.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        Returns
+        -------
+        xp : np.ndarray, shape (n_samples, n_output_features)
+            The matrix of features, where n_output_features is the number of features
+            generated from applying the custom functions to the inputs.
+        """
+        check_is_fitted(self)
+
+        x = check_array(x)
+        print(x.shape)
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/feature_library/SINDyPI_library.py:311]: fixed

File: pysindy/feature_library/SINDyPI_library.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
+from itertools import combinations
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/feature_library/SINDyPI_library.py:1]: fixed

File: pysindy/optimizers/sindymodified_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            normalize=normalize,
+            fit_intercept=fit_intercept,
+            copy_X=copy_X,
+        )
+
+        if max_iter <= 0:
+            raise ValueError("max_iter must be positive")
+        if threshold < 0.0:
+            raise ValueError("threshold must not be negative")
+        if thresholder != "l1":
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/optimizers/sindymodified_optimizer.py:139]: This is actually a mistake on my end. "Modified SINDy" is the name I used for Kardy's neural network version and I didn't think I had added this to the repo. I will delete it now. The optimizer to look at is sindypi_optimizer.py

File: pysindy/pysindy.py (hunk 3 above)
Comment (akaptano) [reply pysindy/pysindy.py:11]: seems to be the standard for python ODEs these days. Originally, I hoped it could solve differential algebraic equations (DAEs) of the sort that SINDy-PI produces. It appears not... need special functionality (see the example notebook). My recommendation would be to change the main branch to use solve_ivp for everything.

File: pysindy/optimizers/sindymodified_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+            raise ValueError("l0 and other nonconvex regularizers are not implemented ")
+        if thresholder[:8].lower() == "weighted" and thresholds is None:
+            raise ValueError("weighted thresholder requires the thresholds parameter")
+        if thresholder[:8].lower() != "weighted" and thresholds is not None:
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/optimizers/sindymodified_optimizer.py:143]: Can you expand on this? I think I am allowing for the weighted_l1 thresholder.

File: pysindy/feature_library/SINDyPI_library.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+                        " number of elements"
+                    )
+            elif (len(library_functions) + len(xdot_library_functions)) != len(
+                function_names
+            ):
+                raise ValueError(
+                    "(x_library_functions + xdot_library_functions) and "
+                    " function_names must have the same"
+                    " number of elements"
+                )
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/feature_library/SINDyPI_library.py:128]: I don't think this works for the case where library_functions is passed, x_dot_library_functions is not passed, and function_names is passed but len(library_functions) != len(function_names), which should throw an error.

**Comment by akaptano** [comment]
Okay, pushing an update with all these fixes and more in a moment. Some more changes include changing the names of TrappingSR3 -> Trapping and SINDyPIoptimizer -> SINDyPI, allowing for a normalize_columns parameter in some of the optimizers (doing the same in the PDEs branch), adding L2 (ridge regression) functionality into SR3 and other branches (although doesn't seem to work very well...), updating the examples to be consistent, and more.

**Comment by akaptano** [comment]
Also, I recommend we rename the STLSQ optimizer to STRidge, since the current implementation really is ridge regression, reducing to least-squares only when alpha = 0. The name currently implies the default is to do pure least-squares.

**Comment by briandesilva** [comment]
> Also, I recommend we rename the STLSQ optimizer to STRidge, since the current implementation really is ridge regression, reducing to least-squares only when alpha = 0. The name currently implies the default is to do pure least-squares.

I see your point, it isn't pure least squares when alpha is nonzero. But there are a few reasons I think we should leave the name as-is:
* A tiny bit of regularization helps stabilize least-squares quite a bit. The default alpha value of 0.05 is small enough that it shouldn't have much of an effect for clean systems, but can help for noisy ones (especially for new users who don't know what they're doing).
* It allows us to avoid creating another optimizer
* To avoid confusing PySINDy's early-adopters
* I think least-squares is a more popular name than ridge regression—it's not uncommon to call ridge regression "regularized least squares"

I think that as long as we're clear that unless you set alpha=0 `STLSQ` is actually doing ridge regression, we should be okay. This probably means adding a bit of text to the `STLSQ` docstring.

**Comment by akaptano** [comment]
Sounds good Brian

**Review by briandesilva — CHANGES_REQUESTED** [review]
Thanks, Alan, for addressing all my earlier comments! I think things are in good shape overall, but there are some loose ends to tie up regarding testing and the examples. There are also some minor comments throughout the code.

### Testing
We still need a lot of unit and integration tests for the new functionality. My advice here is to start by incorporating new optimizers and libraries into existing tests. Sometimes a new object won't fit into the existing tests. In that case we can create similar versions of existing tests for the new object. Next, check the [code coverage report](https://app.codecov.io/gh/dynamicslab/pysindy/compare/119/tree/pysindy) to look for gaps in the existing tests. This report tracks the percentage of our code that is covered by tests. Our coverage should be in the mid 90s.

### Examples
`1_feature_overview`
* This notebook is meant to cover virtually all of PySINDy's functionality, so we need to add either examples of the new optimizers and libraries or add references to the other notebooks where users can find examples (specifically `SINDyPILibrary`, `SINDyPI`, `Trapping`, and `CustomLibrary` with linear control variables).
* I like your idea to move the ODEs into one location. However for this particular notebook I think it's important for the users to see the ODEs that we're attempting to discover. That way they can see that the coefficients PySINDy spits out are accurate. I can see two options here: (1) put the definitions of the equations back in this notebook or (2) write down the mathematical expressions for the equations (like in notebook 2) using tex.
 
`9_sindypi_examples`
* It seems like you're still working on this notebook, so I'll hold off on doing a full review of it for now. That being said, I have a few minor comments.
* We don't lint the notebooks, but you should still try to follow the linting guidelines, import order, etc. in the notebooks. One option is to paste code into a file then run `black` on it.
* Be sure to give lots of explanation of what you're doing and what's going on in the notebooks. We want users to be able to figure out how/when to use different features of the package.

**Comment by akaptano** [comment]
Hey Brian, given that this branch also needs all the scipy version 1.0 changes, I think I'm going to clean it up, merge it into the PDEFIND branch, and just do one giant branch on PDEFIND. Let's do comments there but I'll resolve the ones so far here.

File: pysindy/optimizers/trapping_sr3.py [hunk 10]
```diff
@@ -13,7 +13,7 @@
 from .sr3 import SR3
 
 
-class TrappingSR3(SR3):
+class Trapping(SR3):
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/optimizers/trapping_sr3.py:16]: reverting this change, seems like a pointless change

File: pysindy/optimizers/trapping_sr3.py [hunk 11]
```diff
@@ -397,8 +399,15 @@ def _solve_sparse_relax_and_split(self, r, N, x_expanded, y, Pmatrix, A, coef_pr
         else:
             prob = cp.Problem(cp.Minimize(cost))
 
-        # default solver is OSQP here
+        # default solver is OSQP here but switches to ECOS for L2
+        # try:
```
[the remark this answers is not in the document]
Comment (akaptano) [reply pysindy/optimizers/trapping_sr3.py:403]: I can remove some of it. Unfortunately, depending on the regularizer, cvxpy will use a different optimization method, with its own keyword arguments, so it sometimes throws an error.

File: test/feature_library/test_feature_library.py [hunk 12]
```diff
@@ -13,6 +13,7 @@
 from pysindy.feature_library import FourierLibrary
 from pysindy.feature_library import IdentityLibrary
 from pysindy.feature_library import PolynomialLibrary
+from pysindy.feature_library import SINDyPILibrary
```
[the remark this answers is not in the document]
Comment (akaptano) [reply test/feature_library/test_feature_library.py:16]: Agreed. This is a common issue with the Custom, PDE, and SINDy-PI libraries. The main issue is that the custom library features seem to be tested through pytest.lazy_fixture, but I cannot figure out how the heck to use this. Maybe I'll leave these for you to add.
